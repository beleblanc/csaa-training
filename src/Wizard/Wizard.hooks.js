import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState,
} from "react";

export const WizardContext = createContext();
export const useWizardContext = () => {
  const context = useContext(WizardContext);
  return context;
};

export const useWizard = ({ children, submitHandler }) => {
  const [numberOfPages, setNumberOfPages] = useState();
  const [currentPageNumber, setCurrentPageNumber] = useState(1);

  useEffect(() => {
    console.log(children.length);
    setNumberOfPages(children.length);
    setCurrentPageNumber(1);
  }, [children.length]);

  const goToNextPage = useCallback(() => {
    console.log(currentPageNumber < numberOfPages);
    if (currentPageNumber < numberOfPages) {
      console.log(currentPageNumber);
      setCurrentPageNumber(currentPageNumber + 1);
    }
  }, [numberOfPages, currentPageNumber]);

  const goToPreviousPage = useCallback(() => {
    if (currentPageNumber > 1) setCurrentPageNumber(currentPageNumber - 1);
  }, [currentPageNumber]);

  const goToInitialPage = useCallback(() => {
    setCurrentPageNumber(1);
  }, []);

  const context = {
    currentPageNumber,
    numberOfPages,
    isLastPage: currentPageNumber === numberOfPages,
    goToNextPage,
    goToPreviousPage,
    submitHandler: () => {
      submitHandler();
      goToInitialPage();
    },
  };

  return context;
};
