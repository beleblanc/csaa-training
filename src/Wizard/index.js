export {default, useWizardContext } from './Wizard';