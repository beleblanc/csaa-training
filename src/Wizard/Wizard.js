import React from "react";
import ForwardBtn from "./ForwardButton";
import BackButton from "./BackButton";
import Panel from "./Panel/Panel";
import { WizardContext, useWizard } from "./Wizard.hooks";

const Wizard = (props) => {
  const context = useWizard(props);

  return (
    <WizardContext.Provider value={context}>
      <div className="wizard">
        <div className="wizard-content-area">
          {props.children.map((child, index) => (
            <Panel pageNumber={index + 1} key={`child-of-panel-${index}`}>
              {child}
            </Panel>
          ))}
        </div>
        <div className="wizard-buttons">
          <ForwardBtn /> <BackButton />
        </div>
      </div>
    </WizardContext.Provider>
  );
};

export default Wizard;
