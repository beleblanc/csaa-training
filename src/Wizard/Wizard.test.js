import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import React from "react";
import Wizard from "./Wizard";
const submitHandler = jest.fn();
describe("Wizard", () => {
  it("should render as expected", () => {
    const { container } = render(
      <Wizard submitHandler={submitHandler}>
        <div>Page1</div>
        <div>Page2</div>
        <div>Page3</div>
      </Wizard>
    );

    expect(container).toMatchSnapshot();
  });
  it("should navigate when the buttons are pressed", async () => {
    render(
      <Wizard submitHandler={submitHandler}>
        <div>Page1</div>
        <div>Page2</div>
        <div>Page3</div>
      </Wizard>
    );

    fireEvent.click(screen.getByText("Next"));
    await waitFor(async () =>
      expect(await screen.findByText("Page2")).toBeInTheDocument()
    );
    fireEvent.click(screen.getByText("Next"));
    await waitFor(async () =>
      expect(await screen.findByText("Page3")).toBeInTheDocument()
    );
    fireEvent.click(screen.getByText("Submit"));
    await waitFor(async () => expect(submitHandler).toHaveBeenCalled());
  });
});
