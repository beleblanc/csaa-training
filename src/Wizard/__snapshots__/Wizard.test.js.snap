// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard should render as expected 1`] = `
<div>
  <div
    class="wizard"
  >
    <div
      class="wizard-content-area"
    >
      <div
        class="wizard-panel"
      >
        <div>
          Page1
        </div>
        <div
          class="notification-center"
        >
          You are on page numbers: 
          1
        </div>
      </div>
    </div>
    <div
      class="wizard-buttons"
    >
      <div
        class="fowardBtn_div"
      >
        <button
          type="button"
        >
          Next
        </button>
      </div>
       
    </div>
  </div>
</div>
`;
